docs: First class webhooks (#45352)
sourcegraph/sourcegraph: doc/admin/config/webhooks.md

# Incoming webhooks

Incoming webhooks can be configured on a Sourcegraph instance in order to receive webhook events from code hosts. This allows Sourcegraph to react more quickly to events that occur outside the instance instead of polling for changes.

> Webhooks are currently implemented to speed up two types of external events:

Reviewer: ```suggestion
Webhooks are currently implemented to speed up two operations:
```

Both items in the list aren't external events

* Keeping batch changes changeset details up to date
* Keeping code on Sourcegraph fresh by responding to new code being pushed to a repository

See the table below for code host compatibility:

Code host | [Batch changes](../../batch_changes/index.md) | Code push
--------- | :-: | :-:
GitHub | 🟢 | 🟢 
GitLab | 🟢 | 🔴
Bitbucket Server / Datacenter | 🟢 | 🔴 
Bitbucket Cloud | 🟢 | 🔴

> To receive webhooks both the receiving and sending side need to be configured. To configure the receiving side, [add an incoming webhook](#adding-an-incoming-webhook). Set up the sending side by [configuring webhooks on your code host](#configuring-webhooks-on-the-code-host)

Reviewer: ```suggestion
In order to receive webhooks both the receiving and sending side need to be configured:

1. [Add an incoming webhook](#adding-an-incoming-webhook) to be able to receive webhooks on the Sourcegraph side.
1. Start to receive webhooks by [adding the webhooks to your code host](#configuring-webhooks-on-the-code-host)
```

## Adding an incoming webhook

Before adding an incoming webhook you should ensure that you have at least one [code host connection](../external_services/index.md) configured. 

The incoming webhook will be configured to accept events from a specific code host connection based on its type and URN.

1. Navigate to **Site Admin > Incoming webhooks**
1. Click **Add webhook**
1. Fill out the form:
   1. **Webhook name**: Descriptive name for the webhook.
   1. **Code host type**: Select from the dropdown. This will be filtered based on code host connections added on your instance.
   1. **Code host URN**: The URN for the code host. Again, this will be filtered by code host connections added on your instance.

>    1. **Secret**: An arbitrary shared secret between Sourcegraph and the code host. A default value is provided, but you are free to change it.
>        > NOTE: Secrets are not supported for BitBucket cloud

Reviewer: ```suggestion
   1. **Secret**: An arbitrary shared secret between Sourcegraph and the code host. A default value is provided, but you are free to change it. _Secrets are not supported for Bitbucket Cloud._
```

The `NOTE` doesn't render. It just shows inline and our parser doesn't like inline blocks in our lists, so instead let's just call it out like this.

> 1. Click **Create**
> 
> The incoming webhook will now be created, and you will be redirected to a page showing more details.
> 
> Use the unique URL present on the details page to configure [the webhook on your code host](#configuring-webhooks-on-the-code-host).

Reviewer: ```suggestion
1. Click **Create** to create the webhook and be redirect to a page showing more details
1. Use the unique URL present on the details page to configure [the webhook on your code host](#configuring-webhooks-on-the-code-host).
```

# Webhooks

> Webhook receivers can be configured on a Sourcegraph instance in order to receive webhook events from code hosts. This allows Sourcegraph to react more quickly to events that occur outside of the instance. 

Reviewer: Is "webhook receivers" the official name? In the UI they're called "Incoming webhooks"
<img width="756" alt="screenshot_2022-12-08_07 21 33@2x" src="https://user-images.githubusercontent.com/1185253/206372923-553741c2-04e8-4654-b56c-db6dc62e818f.png">

I think we should be consistent then. I also think maybe this page should be called "Incoming webhooks" instead of just "Webhooks"

We currently support webhooks for the following:

Code host | [Batch changes](../../batch_changes/index.md) | Repository push
--------- | :-: | :-:
GitHub | 🟢 | 🟢 
GitLab | 🟢 | 🔴
Bitbucket Server / Datacenter | 🟢 | 🔴 

> Bitbucket Cloud | 🟢 | 🔴

Reviewer: Is this table helpful at this point in the doc? Does a reader know what "Repository push" means? (I don't)

> Webhooks need to be configured both on the sending side, the code host and receiveing side, Sourcegraph.

Reviewer: ```suggestion
Webhooks need to be configured both on the sending and the recieving side, the code host and Sourcegraph respectively.
```

Reviewer: Shouldn't we link to "get started" here or something? It's a bit weird that this sentence ends with nothing and then next up there's the "Deprecation notice"

Author: I reworded this and linked to the setup section for incoming hooks and code hosts.

## Deprecation notice

As of Sourcegraph 4.3.0 webhooks added via code host configuration are deprecated and support will be removed in release 4.6.0.

This includes any webhooks pointed at URLs starting with the following:

* `.api/github-webhooks`

> * `.api//gitlab-webhooks`

Reviewer 2: Typo:

```suggestion
* `.api/gitlab-webhooks`
```

* `.api/bitbucket-server-webhooks`
* `.api/bitbucket-cloud-webhooks`

> In order to continue using webhooks you need to follow the steps below to [add a receiver](#adding-a-receiver) and then update the webhook configured on your code host with the new webhook url which will look something like `https://sourcegraph-instance/.api/webhooks/{UUID}`

Reviewer: ```suggestion
In order to continue using webhooks you need to follow the steps below to [add a receiver](#adding-a-receiver) and then update the webhook configured on your code host with the new webhook URL which will look something like `https://sourcegraph-instance/.api/webhooks/{UUID}`
```

Do we even need this sentence? It essentially says "In order to continue using webhooks, setup webhooks", which is what the rest of the document explains. I think it can probably be shortened to "If you want to continue using incoming webhooks, read on to learn how to setup a new incoming webhook."

> ## Adding a receiver

Reviewer: Again: I think "Creating an incoming webhook" or "Setting up an incoming webhook" might be better. No reader knows what "a receiver" is. Only the first sentence says "webhook receivers".

Before adding a webhook receiver you should ensure that you have at least one [code host connection](../external_services/index.md) configured. 

> In order to receive webhook events you need to add a receiver. The receiver will be configured to accept events from a specific code host connection based on it's type and URN.

Reviewer: I think the first sentence is redundant.

> 1. Navigate to Site Admin > Incoming webhooks

Reviewer: ```suggestion
1. Navigate to **Site Admin > Incoming webhooks**
```

See the batch changes docs you removed above for how to format UI instructions (I couldn't find anything in the handbook)

> 1. Click `Add webhook`

Reviewer: ```suggestion
1. Click **Add webhook**
```

> 1. You'll be presented with a form to create a new webhook receiver:

Reviewer: The rest of the doc was imperative ("Navigate to", "Click [...]"), so we should be consistent and not switch to passive observations now.

```suggestion
1. Fill out the form:
```
